Add dump and diagnose command

Patryk-Stefanski opened the pull request:
# Add `diagnose` and `dump` commands for Kuadrant troubleshooting

  ## Overview

  This PR introduces two new diagnostic commands to `kuadrantctl` that help users troubleshoot and investigate issues with their Kuadrant and Gateway API resources.
  
 This could also be paired up with a claude / command to help debug/fix the environment.
 
 
  ## Use Cases

  - Troubleshooting: Quickly identify why Gateways aren't programmed or policies aren't being enforced
  - Support Tickets: Export complete resource manifests to share with support teams
  - Documentation: Generate snapshots of working configurations

  ## What's New

  ### 1. `kuadrantctl diagnose` - Health Check Command

  A comprehensive diagnostic tool that analyzes Kuadrant and Gateway API resources to identify configuration issues and unhealthy states.

  **Features:**
  - Scans all Kuadrant and Gateway API resources in your cluster
  - Identifies unprogrammed Gateways, unenforced policies, and configuration problems
  - Provides detailed issue reports with reasons and messages from resource conditions
  - Supports namespace filtering or cluster-wide diagnostics

  **Usage:**
  ```bash
  # Diagnose all namespaces
  kuadrantctl diagnose

  # Diagnose specific namespace
  kuadrantctl diagnose -n my-namespace

  # Explicit all-namespaces flag
  kuadrantctl diagnose -A

  Sample Output:
  ================================================================================
  KUADRANT DIAGNOSTIC REPORT
  ================================================================================

  SUMMARY:
  --------------------------------------------------------------------------------
  Gateways:                   2 total, 1 unprogrammed
  HTTPRoutes:                 5 total, 1 unaccepted
  AuthPolicies:               3 total, 1 unenforced
  RateLimitPolicies:          2 total, 0 unenforced
  DNSPolicies:                1 total, 0 unenforced
  TLSPolicies:                1 total, 0 unenforced
  Kuadrants:                  1 total, 0 unready

  Total issues found: 3

  --------------------------------------------------------------------------------
  GATEWAY ISSUES:
  --------------------------------------------------------------------------------

    Resource: gateway-namespace/my-gateway
    Status:   Not Programmed
    Issues:
      - Not Programmed: NoResources - No backend resources available
      - Listener http - Programmed: Pending - Waiting for IP assignment

  --------------------------------------------------------------------------------
  HTTPROUTE ISSUES:
  --------------------------------------------------------------------------------

    Resource: app-namespace/api-route
    Status:   Not Accepted
    Issues:
      - Not Accepted by parent my-gateway: NoMatchingListenerHostname - HTTPRoute hostname doesnt match any Gateway listener

  --------------------------------------------------------------------------------
  AUTHPOLICY ISSUES:
  --------------------------------------------------------------------------------

    Resource: app-namespace/api-auth
    Status:   Not Enforced
    Issues:
      - Not Enforced: TargetNotFound - Target HTTPRoute not found in namespace

  ================================================================================
```

  2. kuadrantctl dump - Resource Export Command

  Exports all Kuadrant and Gateway API resources to files for offline investigation, debugging, or sharing with support teams.

  Features:
  - Dumps all Kuadrant resources (Gateways, HTTPRoutes, AuthPolicies, RateLimitPolicies, DNSPolicies, TLSPolicies, Kuadrants)
  - Organizes resources by type in separate directories
  - Removes cluster-specific metadata (UIDs, resource versions, managed fields) for cleaner output
  - Supports YAML and JSON output formats
  - Creates timestamped directories automatically

  Usage:
  ```
  # Dump all resources to timestamped directory
  kuadrantctl dump

  # Dump to specific directory
  kuadrantctl dump -o my-debug-folder

  # Dump specific namespace
  kuadrantctl dump -n my-namespace

  # Dump in JSON format
  kuadrantctl dump --format json

  Sample Output:
  Resources dumped to: kuadrant-dump-20250102-143055

  Directory Structure:
  kuadrant-dump-20250102-143055/
  ├── gateways/
  │   ├── gateway-namespace-my-gateway.yaml
  │   └── istio-system-istio-gateway.yaml
  ├── gatewayclasses/
  │   └── istio.yaml
  ├── httproutes/
  │   ├── app-namespace-api-route.yaml
  │   └── app-namespace-web-route.yaml
  ├── authpolicies/
  │   └── app-namespace-api-auth.yaml
  ├── ratelimitpolicies/
  │   └── app-namespace-api-ratelimit.yaml
  ├── dnspolicies/
  │   └── gateway-namespace-dns-policy.yaml
  ├── tlspolicies/
  │   └── gateway-namespace-tls-policy.yaml
  └── kuadrants/
      └── kuadrant-system-kuadrant.yaml
```

  How It Works Under the Hood

  Both commands leverage the controller-runtime client to interact with the Kubernetes API:


  1. diagnose command:
    - Inspects status conditions on each resource (e.g., Programmed, Accepted, Enforced)
    - Identifies resources where condition status is not True
    - Aggregates issues and generates summary statistics
    - Outputs formatted report to stdout
  4. dump command:
    - Retrieves full resource manifests from the cluster
    - Strips cluster-specific metadata (UID, resourceVersion, managedFields)
    - Marshals to YAML/JSON
    - Writes each resource to a separate file organized by type

Patryk-Stefanski committed:
Add dump and diagnose command

Signed-off-by: Patryk Stefanski <pstefans@redhat.com>

eguzki reviewed (approved):
Sweet 🍬 

LGTM

eguzki commented:
Add some doc?

Boomatang commented on cmd/dump.go:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	// Create output directory
+	if err := os.MkdirAll(dumpOutputDir, 0755); err != nil {
+		return fmt.Errorf("failed to create output directory: %w", err)
```
In `dumpResourceType` you call `os.MkdirAll` for the resource. The path used is for build from the `dumpOutputDir` and the resource name. Why do we need to make this call here? There is a number of possible errors that can happen after this which could cause blank directories to be created.

Patryk-Stefanski replied:
We don't need to make it here. I'll refactor the code to move the directory creation to just before the file write, ensuring we only create directories when we actually have data to write.

Boomatang commented on cmd/dump.go:
```diff
@@ -0,0 +1,285 @@
+
+	cmd.Flags().StringVarP(&dumpNamespace, "namespace", "n", "", "Namespace to dump resources from (default: all namespaces)")
+	cmd.Flags().BoolVarP(&dumpAllNamespaces, "all-namespaces", "A", false, "Dump resources from all namespaces")
+	cmd.Flags().StringVarP(&dumpOutputDir, "output", "o", "", "Output directory (default: ./kuadrant-dump-<timestamp>)")
```
On allowing the user to set the output directory. What happens if the directory is not empty? Are we overriding existing data in that directory?

Patryk-Stefanski replied:
Yep, it's overwriting the existing data in that directory. Ill add a --force flag and check if the directory is non-empty before proceeding

Boomatang commented on cmd/dump.go:
```diff
@@ -0,0 +1,285 @@
+	k8sClient client.Client,
+	resourceType string,
+	listObj client.ObjectList,
+	itemsFunc func(client.ObjectList) []runtime.Object,
```
Why not pass in the struct is being looped over?

Patryk-Stefanski replied:
I'll refactor the code to use a generic approach.

Boomatang commented on cmd/dump.go:
```diff
@@ -0,0 +1,285 @@
+		obj.SetManagedFields(nil)
+		obj.SetResourceVersion("")
+		obj.SetUID("")
+		obj.SetGeneration(0)
```
Why are you clearing these? If I am dumping resources, I would want to be able to check the resource version and generation against other points of information. The UID maybe I don't want that, but by and large I would prefer to have it.

Boomatang commented on cmd/dump.go:
```diff
@@ -0,0 +1,285 @@
+		} else {
+			// Default to YAML
+			data, err = yaml.Marshal(item)
+		}
```
Why have this check if there is no difference?

Patryk-Stefanski replied:
some stale code, that shouldve been removed

Boomatang commented on cmd/dump.go:
```diff
@@ -0,0 +1,285 @@
+	}
+
+	cmd.Flags().StringVarP(&dumpNamespace, "namespace", "n", "", "Namespace to dump resources from (default: all namespaces)")
+	cmd.Flags().BoolVarP(&dumpAllNamespaces, "all-namespaces", "A", false, "Dump resources from all namespaces")
```
Can there be some input validation on the inputs. The user should not be able to set the namespace and all-namespaces at the same time. 

I also don't think the default should be to dump all the namespaces if not set.

Patryk-Stefanski replied:
Ill add some validation , however the namesapce will be defaulted to all as we cant assume that all users will install their resources in a specific namespace and dumping the required resources from all namespaces is more convenient.

Boomatang replied:
If we can't set the default namespace from the kubeconfig, we should force the setting of the field. I understand it maybe more convenient to have it set to all namespaces, but that is not predictable. Take for example, the user runs `kubectl get kuadrant`, what output would you expect.

Patryk-Stefanski replied:
I would expect the kuadrant resource from whatever namespace the kubeconfig is currently set to. And that makes sense when you're only retrieving one resource that is in a specific namespace. 

When retrieving multiple resources that could be in different namespaces, I think it makes sense to dump resources from all namespaces by default, as this is more than likely what most users will want to do. 

@eguzki wdyt ? default to specific namespace or all-namespaces?

eguzki replied:
As long as the desired behavior is documented, both approaches LGTM. I guess I am trying to say pick the easiest to implement and make it clear in doc.

Boomatang commented on cmd/diagnose.go:
```diff
@@ -0,0 +1,678 @@
+		RunE:  runDiagnose,
+	}
+
+	cmd.Flags().StringVarP(&diagnoseNamespace, "namespace", "n", "", "Namespace to diagnose (default: all namespaces)")
+	cmd.Flags().BoolVarP(&diagnoseAllNamespaces, "all-namespaces", "A", false, "Diagnose resources from all namespaces")
```
Same comment on validation and default as before.

Boomatang commented on cmd/diagnose.go:
```diff
@@ -0,0 +1,678 @@
+	k8sClient, err := client.New(cfg, client.Options{Scheme: scheme.Scheme})
+	if err != nil {
+		return fmt.Errorf("failed to create Kubernetes client: %w", err)
+	}
```
This looks to be the same in the other command. Can this be refactored to be a shared command.

Boomatang commented on cmd/diagnose.go:
```diff
@@ -0,0 +1,678 @@
+				}
+			case "Enforced":
+				if string(condition.Status) != string(metav1.ConditionTrue) {
+					issues = append(issues, fmt.Sprintf("Not Enforced: %s - %s", condition.Reason, condition.Message))
+					report.Summary.UnenforcedRateLimitPolicies++
+				} else {
+					isEnforced = true
+				}
+			}
+		}
```
These blocks seem to be repeated in all the diagnose functions, can it be refactored to be more general?

Boomatang commented on cmd/diagnose.go:
```diff
@@ -0,0 +1,678 @@
+	}
+
+	fmt.Println("\n" + strings.Repeat("=", 80))
+}
```
Can this be converted in to some loop?

Boomatang reviewed (changes requested):
There is some changes I would like to see.

- flag validation
- namespace required explicitly, not defaulting to all, if anything the default should be from the kubeconfig. Like what the kubectl does.
-  Created directories only when ready to write data to location.

Patryk-Stefanski commented:
@Boomatang fyi I'm putting this PR on hold. I want to chat with Chris and the wider team if it's worth keeping and maintaining this repo, or at the very least removing some commands if we're not planning on supporting them.

Boomatang commented:
> if it's worth keeping and maintaining this repo

Can you share links to where these conversations are going on? In the dns-operator we are making design for kuadrant-dns plugin based on this repo existing.

Patryk-Stefanski commented:
It will be raised and discussed at the next kudarant community call

Patryk-Stefanski committed:
remove make generate commands

Signed-off-by: Patryk Stefanski <pstefans@redhat.com>

Patryk-Stefanski commented:
@Boomatang this is ready for review again, I addressed your comments and removed the generate commands as the outcome of the discussion on  this weeks community call was that no one had hard opinions/thoughts on keeping those commands in.

Boomatang commented:
Was there any feedback from Chris on the future of those commands? During the community call it was mentioned that you would be speaking with them during the week.

Boomatang commented:
Also can you split this in to two PRs. The adding of one function has nothing to do with the removal of the others.

Patryk-Stefanski commented:
@Boomatang was not planning on speaking with Chris about this, when you say those commands do you mean the new ones Im adding in or the generate commands Im removing ?

Patryk-Stefanski commented:
> Also can you split this in to two PRs. The adding of one function has nothing to do with the removal of the others.


New commands require the most recent kuadrant api. The generate commands rely on the outdated api, since this PR is moving to use the most up to date api. I think it makes sense to remove the commands in this PR as otherwise there will exist a version in which those commands are broken.

Patryk-Stefanski commented:
Closing this for now, as it's gotten stale, and with similar results can be achieved with the help of  Claude
